### PR TITLE
[server] OSS Allowlist

### LINKED ISSUE #6918: Allow whitelisting github profiles for prof open-source
We'd like to proactively whitelist open-source developers, so that when they sign up they are already on the professional open-source plan. The list would probably grow to max 50k entries. It should be enough to update the list through code PRs and deployments. So can be either DB migrations or in-memory (haven't done the math). This should also work for existing sign-ups so I guess we'd basically need to check the list whenever a user logs in.

cc @ghuntley please add context if you like.

### PR DESCRIPTION
**Pull request by geropl**
## Related Issue(s)

Fixes #6918

## How to test
 - `kubectl port-forward mysql-0 3306`
 - `mysql -u gitpod -h 127.0.0.1 -p`
 - `INSERT INTO d_b_oss_whitelist SET identity = '<str>'`, example: `github.com/geropl`
 - go to https://gpl-6918-oss-positivelist.staging.gitpod-dev.com/plans
 - expect to see: 
image

## Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **geropl** after the PR has been reviewed.

Associated issue: *#6918*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/ee/payment-endpoint/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/payment-endpoint/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/ee/payment-endpoint/src/accounting/subscription-service.ts
Comment (jankeromnes) [inline components/ee/payment-endpoint/src/accounting/subscription-service.ts:106]: I think there is no need to check for other (paid) subscriptions. It's actually fine to have both (Professional Open Source + a paid plan), see for example my plans:

Screenshot 2021-12-14 at 08 52 48

Having both allows to benefit from the Professional Open Source "status" even if you already pay for another plan.

File: components/gitpod-db/src/typeorm/migration/1638973332787-OssWhitelist.ts
Comment (jankeromnes) [inline components/gitpod-db/src/typeorm/migration/1638973332787-OssWhitelist.ts:18]: Thought: While semantically correct, maybe it's not necessary to delete this table on rollback?

File: components/ee/payment-endpoint/src/accounting/subscription-service.ts
Comment (geropl) [reply components/ee/payment-endpoint/src/accounting/subscription-service.ts:106]: @jankeromnes Do we display that independently? That's news to me! And the screenshot above shows a "Team Subscription". Do we handle/display OSS the same, really? 
I wanted to avoid situations where we "override" peoples paid subscription, and they end up not noticing because of this.

File: components/gitpod-db/src/typeorm/migration/1638973332787-OssWhitelist.ts
Comment (geropl) [reply components/gitpod-db/src/typeorm/migration/1638973332787-OssWhitelist.ts:18]: I like to be clean here. But true, we're lagging general guidelines on how to implement rollback... and we never use it anyway. :see_no_evil:

File: components/ee/payment-endpoint/src/accounting/subscription-service.ts
Comment (jankeromnes) [reply components/ee/payment-endpoint/src/accounting/subscription-service.ts:106]: Yes, we display them independently.

- Professional Open Source is a sort of "status" (if you have it, it replaces your "free plan" card on the left -- as in my screenshot)
- The screenshot above shows a "Team Subscription", because I'm part of a "Team Unleashed" plan -- but that's actually not relevant (if I were on the regular "Unleashed" plan, it would show that as my active plan, without a mention of a team plan, and I'd still have "Professional Open Source" as my left-most card)

**Comment by AlexTugarev** [comment]
@geropl and @ghuntley, while this PR looks good to me, I'm not quite sure this meets the expectations from the initial GH issue (see [comment](https://github.com/gitpod-io/gitpod/issues/6918#issuecomment-986327607).) the open question to me is, if the update to the whitelist should automatically update the subscriptions for existing users.

File: components/ee/payment-endpoint/src/accounting/subscription-service.ts
Comment (geropl) [reply components/ee/payment-endpoint/src/accounting/subscription-service.ts:106]: Removed the paid-plan check.

I kept the check for "Prof. OSS", though, to not cancel/add a new subscription on each login.

**Comment by geropl** [comment]
> if the update to the whitelist should automatically update the subscriptions for existing users.

@AlexTugarev This is answered in the [original issue comment](https://github.com/gitpod-io/gitpod/issues/6918#issue-1064216107): 

> This should also work for existing sign-ups so I guess we'd basically need to check the list whenever a user logs in.

That's how it's implemented [here](https://github.com/gitpod-io/gitpod/pull/7215/files#diff-2100cdef3892d760e003e047dbccdbf053f5c6456d4c239af1b1f7e58f2baa51R89-R92).

**Review by jankeromnes — APPROVED** [review]
Works like a charm! ✨ Many thanks @geropl 🙏

I guess this could do nicely with an admin feature to add users to the allowlist. ✏️

Also, on a general note, I personally prefer these alternative names:
- whitelist → allowlist
- blacklist → blocklist (or denylist)

(Because the white = good, black = bad image is sort of problematic 😬 but I know we have these in other places in the code base, so this could definitely be a follow-up fix. 👍)

Approving because code looks good to me & works as intended, holding due to minor comments in-line. Please take a look, but feel free to decide between fix / ignore / follow-up.

/lgtm
/approve
/hold

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 93b5186fd372780a682cd1bb6f2fafd26f6d54df</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7215#pullrequestreview-831228593" title="Approved">jankeromnes</a>*

Associated issue: *#6918*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/ee/payment-endpoint/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/payment-endpoint/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

**Comment by jankeromnes** [comment]
Many thanks for addressing my comments!

Newer changes still look good to me. 👍 🚀

/lgtm

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: d17164f01310a61a06a4543e719218873c7c2efd</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7215#issuecomment-993565036" title="LGTM">jankeromnes</a>*

Associated issue: *#6918*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/ee/payment-endpoint/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/payment-endpoint/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
